src/vm/runners/cairo_runner.rs

```diff
@@ -756,23 +756,27 @@ impl CairoRunner {
     ///Relocates the VM's trace, turning relocatable registers to numbered ones
     fn relocate_trace(
         &mut self,
-        vm: &mut VirtualMachine,
-        relocation_table: &Vec<usize>,
+        vm: &VirtualMachine,
+        relocation_table: &[usize],
     ) -> Result<(), TraceError> {
         if self.relocated_trace.is_some() {
             return Err(TraceError::AlreadyRelocated);
         }
-
-        let trace = vm.trace.as_ref().ok_or(TraceError::TraceNotEnabled)?.iter();
-        let mut relocated_trace = Vec::<RelocatedTraceEntry>::with_capacity(trace.len());
-        for entry in trace {
-            relocated_trace.push(RelocatedTraceEntry {
-                pc: relocate_trace_register(entry.pc, relocation_table)?,
-                ap: relocate_trace_register(entry.ap, relocation_table)?,
-                fp: relocate_trace_register(entry.fp, relocation_table)?,
-            })
-        }
-        self.relocated_trace = Some(relocated_trace);
+        let segment_1_base = relocation_table
+            .get(1)
+            .ok_or(TraceError::NoRelocationFound)?;
+        self.relocated_trace = Some(
+            vm.trace
+                .as_ref()
+                .ok_or(TraceError::TraceNotEnabled)?
+                .iter()
+                .map(|entry| RelocatedTraceEntry {
+                    pc: entry.pc_off + 1,
+                    ap: entry.ap_off + segment_1_base,
+                    fp: entry.fp_off + segment_1_base,
+                })
+                .collect(),
+        );
         Ok(())
     }
```

Comment:
Now that it's `[usize]` we could probably do the relocation in-place.

Reply:
Added in-place relocation [524413d)